### PR TITLE
Add dest mw take two

### PR DESCRIPTION
**Pull request by aultimus**
This PR adds destination middleware support. This feature was originally released to master but was then reverted as a regression was discovered. That regression was addressed by this commit ( https://github.com/segmentio/analytics.js-core/pull/150/commits/010219b675e859d35deafaf8c1e387461839c694 ) and we are re-releasing.

## Verification

Run this code in the console of a project connected to a mixpanel destination and then observe the identify event in mixpanel with user id foobarcat

```
f=function({payload, next, integrations}) {payload.obj.userId = "foo"; next(payload);}; analytics.addSourceMiddleware(f);
f=function(payload, integration, next) {payload.obj.userId = payload.obj.userId + "bar";next(payload);}; analytics.addIntegrationMiddleware(f);
f=function({payload, integration, next}) { payload.obj.userId = payload.obj.userId + "cat";next(payload);}; analytics.addDestinationMiddleware("Mixpanel",[f]);
analytics.identify();
```

See:
https://www.dropbox.com/s/ncrbwm3tkvphhzf/Screen%20Recording%202020-05-21%20at%202.26.55%20PM.mov?dl=0

## Checklist

Please ensure the following are completed to help get your PR merged:

- [x] Thorough explanation of the issue/solution, and a link to the related issue
- [x] CI tests are passing
- [ ] Unit tests were written for any new code
- [ ] Code coverage is at least maintained, or increased.

## Respect earns Respect 👏

Please respect our Code of Conduct, in short:

- Using welcoming and inclusive language.
- Being respectful of differing viewpoints and experiences.
- Gracefully accepting constructive criticism.
- Focusing on what is best for the community.
- Showing empathy towards other community members.

### CODE REVIEW

File: lib/analytics.js [hunk 1]
```diff
@@ -791,12 +821,43 @@ Analytics.prototype._invoke = function(method, facade) {
...
-                method: method,
-                integration_name: integration.name
-              });
+              // apply destination middlewares
+              // Apply any integration middlewares that exist, then invoke the integration with the result.
+              if (self._destinationMiddlewares[integration.name]) {
+                self._destinationMiddlewares[integration.name].applyMiddlewares(
+                  facadeCopy,
+                  integration.name,
+                  function(result) {
```
Comment (juliofarah) [inline lib/analytics.js:830]: is there a way, using JSDocs, document the format we expect `result` to respect?

File: lib/analytics.js [hunk 2]
```diff
@@ -791,12 +821,43 @@ Analytics.prototype._invoke = function(method, facade) {
...
+                    // A nullified payload should not be sent to an integration.
+                    if (result === null) {
```
Comment (juliofarah) [inline lib/analytics.js:832]: is a nullified payload something unexpected that we should be worried about? If so, should we report metrics on them?

Comment (aultimus) [reply lib/analytics.js:832]: I think this is the case where the user explicitly wants to drop the event, so its a feature really

**Comment by juliofarah** [comment]
The code and PR documentation look good to me!
As I don't have a lot of context on AJS as a whole, a newbie question I have is: although this PR is pretty well covered in terms of unit tests, how do we treat end-to-end and integration testing when implementing features like this?

File: lib/analytics.js (hunk 2 above)
Comment (juliofarah) [reply lib/analytics.js:832]: gotcha! thanks for the clarification :D

File: lib/analytics.js (hunk 1 above)
Comment (aultimus) [reply lib/analytics.js:830]: I think you would need to do one line per property like this https://jsdoc.app/tags-param.html#parameters-with-properties

File: HISTORY.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+- Revert "Add destination mw (#148)"
```
Comment (pooyaj) [inline HISTORY.md:1]: Let's remove the "Revert" line, since it will be i the release log, and confuses people

File: test/analytics.test.js [hunk 4]
```diff
@@ -495,7 +495,7 @@ describe('Analytics', function() {
     });
 
     it('should emit "invoke" with facade', function(done) {
-      var opts = { All: false };
+      var opts = { All: true };
```
Comment (pooyaj) [inline test/analytics.test.js:498]: can we revert this change. I wonder if we need this change even after fixing the `emitter` issue.

Comment (aultimus) [reply test/analytics.test.js]: resolved by https://github.com/segmentio/analytics.js-core/pull/150/commits/fccfece9c33838866a52cb792e85cc3f3060d51b

File: HISTORY.md (hunk 3 above)
Comment (aultimus) [reply HISTORY.md]: resolved by https://github.com/segmentio/analytics.js-core/pull/150/commits/fccfece9c33838866a52cb792e85cc3f3060d51b
